Parse types structurally

martint opened the pull request.

martint committed:
Parse types structurally

Types are now fully unrolled into syntax tree nodes.

dain commented on presto-parser/src/main/java/io/prestosql/sql/tree/DateTimeDataType.java:
```diff
@@ -0,0 +1,91 @@
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        io.prestosql.sql.tree.DateTimeDataType that = (io.prestosql.sql.tree.DateTimeDataType) o;
```
qualified import?

dain commented on presto-parser/src/main/java/io/prestosql/sql/tree/GenericDataType.java:
```diff
@@ -0,0 +1,81 @@
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        io.prestosql.sql.tree.GenericDataType that = (io.prestosql.sql.tree.GenericDataType) o;
```
qualified import?

dain commented on presto-parser/src/main/java/io/prestosql/sql/tree/RowDataType.java:
```diff
@@ -0,0 +1,137 @@
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        io.prestosql.sql.tree.RowDataType that = (io.prestosql.sql.tree.RowDataType) o;
```
qualified import?

dain commented on presto-parser/src/main/java/io/prestosql/sql/tree/RowDataType.java:
```diff
@@ -0,0 +1,137 @@
+        @Override
+        public String toString()
+        {
+            return name + " " + type;
```
I believe this will render with `Optional`.  Is that what you want?

martint replied:
It's for debugging, so it doesn't really matter. But I'll update it to omit the name if missing.

dain commented on presto-parser/src/test/java/io/prestosql/sql/parser/ParserAssert.java:
```diff
@@ -0,0 +1,88 @@
+
+    public static AssertProvider<ParserAssert> type(String sql)
+    {
+        return createAssertion(x -> new SqlParser().createType(x), sql);
```
rename `x`?

dain commented on presto-parser/src/test/java/io/prestosql/sql/parser/TestTypeParser.java:
```diff
@@ -0,0 +1,311 @@
+    @Test
+    public void testComplexTypes()
+    {
+        assertThat(type("ROW(x BIGINT, y DOUBLE, z ROW(m array<bigint>,n map<double,varchar>))"))
```
Maybe make this `DOUBLE PRECISION`?

martint committed:
Clean up formatting of identifiers

There was some code duplication and legacy logic
that was unaware of the fact that Identifier now
carries information about whether the identifier
is delimited.
